crates/aggchain-proof-service/src/service.rs

```diff
@@ -209,3 +246,34 @@ impl tower::Service<AggchainProofServiceRequest> for AggchainProofService {
         .boxed()
     }
 }
+
+type AggchainProofServiceFuture =
+    Pin<Box<dyn Future<Output = Result<AggchainProofServiceResponse, Error>> + Send>>;
+
+impl tower::Service<AggchainProofServiceRequest> for AggchainProofService {
```

Comment:
Wold the following work? Instead of having one `tower::Service` impl for both modes with runtime dispatch, have two, `tower::Service<FullAggchainProofServiceRequest>` and `tower::Service<OptimisitcAggchainProofServiceRequest>`. That would allow to have a specific return type for each request type too.

Reply:
But the return type for both "mode" is the same tho

Reply:
> But it means that you'll need to have multiple instance of this service on the upper function.

Ok, in that case, my suggestion does not make that much sense.